docs/design/proposals/rbd-volume-healer.md

```diff
@@ -0,0 +1,102 @@
+  rbd-nbd has to work with NBD module closely, yes! but remember it's
+  mostly unchanged much from years and consider it to be much stable now.
+  NBD is a generic driver and there will be many eyes on it compared to the
+  rbd driver. rbd driver is vendor-specific logic that might also need
+  changes every now and then when a new feature update is available at Ceph.
+
+### Side effects of userspace mounters
+
+Since the rbd-nbd processes run per volume map on the client side i.e.
+inside the `csi-rbdplugin` node plugin, a restart of the node plugin will
+terminate all the rbd-nbd processes, and there is no way to restore
+these processes back to life currently, which could lead to IO errors
+on all the application pods.
+
+![csi-plugin-restart](./images/csi-plugin-restart.svg)
+
+This is where the Volume healer could help.
+
+## Volume healer
+
+Volume healer runs on the start of rbd node plugin and runs within the
+node plugin driver context.
+
+Volume healer does the below,
+
+- Get the Volume attachment list for the current node where it is running
+- Filter the volume attachments list through matching driver name and
+  status attached
+- For each volume attachment get the respective PV information and check
+  the criteria of PV Bound, mounter type
+- Build the StagingPath where rbd images PVC is mounted, based on the
+  KUBELET path and PV object
+- Construct the NodeStageVolume() request and send Request to CSI Driver.
+- The NodeStageVolume() has a way to identify calls received from the
+  healer and when executed from the healer context, it just runs in the
+  minimal required form, where it fetches the previously mapped device to
+  the image, and the respective secrets and finally ensures to bringup the
+  respective process back to life. Thus enabling IO to continue.
+
+### More thoughts
+
+- Currently the NodeStageVolume() call is safeguarded by the global Ceph-CSI
+  level lock (per volID) that needs to be acquired before doing any of the
+  NodeStage, NodeUnstage, NodePublish, NodeUnPulish operations. Hence none
+  of the operations happen in parallel.
+- Any issues if the NodeUnstage is issued by kubelet?
+  - This can not be a problem as we take a lock at the Ceph-CSI level
+  - If the NodeUnstage success, Ceph-CSI will return StagingPath not found
+    error, we can then skip
+  - If the NodeUnstage fails with an operation already going on, in the
+    next NodeUnstage the volume gets unmounted
+- What if the PVC is deleted?
+  - If the PVC is deleted, the volume attachment list might already got
+    refreshed and entry will be skipped/deleted at the healer.
+  - For any reason, If the request bails out with Error NotFound, skip the
+    PVC, assuming it might have deleted or the NodeUnstage might have
+    already happened.
+- The Volume healer currently works with rbd-nbd, but the design can
+  accomidate other userspace mounters (may be cepfs fuse).
```

Comment:
may be good to have a note on its support level too.

Reply:
cepfs fuse/ cephfs fuse

Reply:
We have https://github.com/ceph/ceph-csi/blob/devel/docs/rbd-nbd.md to discuss on the support matrix.

Reply:
I also think that the design doc is not a good fit for support level details.

Reply:
yeah , above support matrix is good enough.